src/libraries/System.IO.FileSystem/src/System/IO/Enumeration/FileSystemEntry.Unix.cs

```diff
@@ -34,52 +33,42 @@ namespace System.IO.Enumeration
+            entry._status.InvalidateCaches();
+
+            bool isDirectory = directoryEntry.InodeType == Interop.Sys.NodeType.DT_DIR;
+            bool isSymlink   = directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK;
+            bool isUnknown   = directoryEntry.InodeType == Interop.Sys.NodeType.DT_UNKNOWN;
 
-            bool isDirectory = false;
-            bool isSymlink = false;
-            if (directoryEntry.InodeType == Interop.Sys.NodeType.DT_DIR)
-            {
-                // We know it's a directory.
-                isDirectory = true;
-            }
             // Some operating systems don't have the inode type in the dirent structure,
             // so we use DT_UNKNOWN as a sentinel value. As such, check if the dirent is a
-            // directory.
-            else if ((directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK
-                || directoryEntry.InodeType == Interop.Sys.NodeType.DT_UNKNOWN)
-                && Interop.Sys.Stat(entry.FullPath, out Interop.Sys.FileStatus targetStatus) >= 0)
+            // symlink or a directory.
+            if (isUnknown)
             {
-                // Symlink or unknown: Stat to it to see if we can resolve it to a directory.
-                isDirectory = (targetStatus.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFDIR;
+                isSymlink = entry._status.IsSymbolicLink(entry.FullPath);
+                isDirectory = entry._status.IsDirectory(entry.FullPath);
             }
+
             // Same idea as the directory check, just repeated for (and tweaked due to the
             // nature of) symlinks.
-            if (directoryEntry.InodeType == Interop.Sys.NodeType.DT_LNK)
-            {
-                isSymlink = true;
-            }
-            else if ((directoryEntry.InodeType == Interop.Sys.NodeType.DT_UNKNOWN)
-                && (Interop.Sys.LStat(entry.FullPath, out Interop.Sys.FileStatus linkTargetStatus) >= 0))
+            // Whether we had the dirent structure or not, we treat a symlink to a directory as a directory,
+            // so we need to reflect that in our isDirectory variable.
+            if (isSymlink)
             {
-                isSymlink = (linkTargetStatus.Mode & Interop.Sys.FileTypes.S_IFMT) == Interop.Sys.FileTypes.S_IFLNK;
+                isDirectory = entry._status.IsDirectory(entry.FullPath);
             }
 
-            entry._status = default;
-            FileStatus.Initialize(ref entry._status, isDirectory);
+            entry._status.InitiallyDirectory = isDirectory;
 
             FileAttributes attributes = default;
             if (isSymlink)
                 attributes |= FileAttributes.ReparsePoint;
             if (isDirectory)
                 attributes |= FileAttributes.Directory;
-            if (directoryEntry.Name[0] == '.')
+            if ((directoryEntry.NameLength > 0 && directoryEntry.Name[0] == '.') || entry._status.HasHiddenFlag) // soft retrieval
```

Comment:
Consider using `IsNameHidden` helper from FileStatus when you address https://github.com/dotnet/runtime/issues/52456.

Reply:
No, here I can't, for two reasons:

1.  `directoryEntry.Name` is not a `ReadOnlySpan<char>`, but a `byte*`. Which is why its length is specified in a separate variable, `NameLength`.

2. If I wanted to access the fileName as a `ROS<char>`, then I would have to initialize the `FileName` property, which would cause a perf hit.

Reply:
Added this as a potential improvement: https://github.com/dotnet/runtime/issues/52456

Reply:
> If I wanted to access the fileName as a ROS<char>, then I would have to initialize the FileName property, which would cause a perf hit.

Not sure if understand this statement but you can do something similar to the following:
`IsNameHidden(new ReadOnlySpan<byte>(directoryEntry.Name, directoryEntry.NameLength))` 
and maybe try adding `MethodImpl.AggresiveInlining` to IsNameHidden.

Reply:
Update: This might not be possible after all. Handling the `byte*` inside the `ROS` might not be the correct thing to do. When I did this, all the unit tests failed, and I think `Name` is not retrieving the full name of the file (up to the size specified by `NameLength`). This might explain why inside the `FileName` property, we are doing an extra p/invoke to `GetName`, to ensure we get the full name of the expected size.
I'll leave this as is.